woocommerce/woocommerce
# Implement `maxFailures` and warning messages

### All Submissions:

### Changes proposed in this Pull Request:

Currently, E2E test runs that have too many failures generate huge Allure test reports, and take a very long time to complete like so:

![Screen Shot 2022-07-19 at 8 23 47 PM](https://user-images.githubusercontent.com/4509348/180435942-d8167e45-db38-4f7e-93cf-c039b089f91c.png)

To address this, we can use Playwright's [`maxFailures`](https://playwright.dev/docs/api/class-testconfig#test-config-max-failures) configuration to implement a fail-fast mechanism to save time on testing a faulty build, and to prevent the Allure reports from getting huge.

Closes https://github.com/woocommerce/woocommerce/issues/34044.

### How to test the changes in this Pull Request:

1. In this pull request or in a duplicate one, edit a few Playwright E2E tests to produce more than 3 failures. For example, you may edit the locator in this test to a wrong one (say, from `#woocommerce-layout__primary` to `#wrong_locator`) so that it would produce more than 3 failures:
    https://github.com/woocommerce/woocommerce/blob/049c0d17cc25179cc0581311aec827897940f429/plugins/woocommerce/tests/e2e-pw/tests/admin-analytics/analytics.spec.js#L29
2. Push this change and wait for the "Run tests against PR" workflow to finish.
3. Verify that the Test Results Summary shows:
    - A "Warning" section that says there were:
        - failed and broken tests
        - incomplete E2E test run
    - The table shows the "Total" for E2E tests is less than 186 (the total number of E2E tests we have currently)
    - The table shows the duration of E2E tests to be just several minutes, not reaching almost an hour.
    - The `e2e-test-report---pr-33977` artifact is only around 50 MB, no longer reaching hundreds of MB's like before.
    <img width="720" alt="Screen Shot 2022-07-22 at 8 29 28 PM" src="https://user-images.githubusercontent.com/4509348/180440595-d664992a-7941-4dcb-ba5c-4193e8eb5f0b.png">
5. Revert the last change to remove the intentional failures we added.
6. This time, edit a few tests so that only 1 or 2 tests would fail. For example, you may edit the expected title in this line so that the E2E test run will produce only 1 failure.
    https://github.com/woocommerce/woocommerce/blob/e2e/pr-fail-fast/plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js#L32
7. Push this change and wait again for the " Run tests against PR" workflow to complete.
8. Verify that the Test Results Summary shows only 1 warning: the one about failures and broken tests. It no longer shows the warning about incomplete E2E test run.
9. Verify that the table shows the "Total" for E2E tests to be 186, which means that all E2E tests were executed.
    <img width="600" alt="Screen Shot 2022-07-22 at 7 51 13 PM" src="https://user-images.githubusercontent.com/4509348/180441935-e3c0ac7f-d952-4599-b2fb-cb8c9f8d6bb4.png">

11. Revert the previous change so that no more failures will occur.
12. Push this change and wait again for the " Run tests against PR" workflow to complete.
13. Verify that all E2E tests ran (186 currently) based on the table in the Test Results Summary.
14. Verify that no more warnings were displayed in the Test Results Summary.
     <img width="600" alt="Screen Shot 2022-07-22 at 8 46 50 PM" src="https://user-images.githubusercontent.com/4509348/180442174-3c806769-bdcf-4670-8332-0a93c952d473.png">


### Other information:


### FOR PR REVIEWER ONLY:
